crates/uv-redacted/src/lib.rs

```diff
@@ -1,21 +1,241 @@
+        let _ = self.0.set_password(None);
+    }
+
+    /// Returns string representation without masking credentials.
+    #[inline]
+    pub fn to_string_with_credentials(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+impl Deref for DisplaySafeUrl {
+    type Target = Url;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for DisplaySafeUrl {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl std::fmt::Display for DisplaySafeUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt_with_obfuscated_credentials(&self.0, f)
+    }
+}
+
+impl std::fmt::Debug for DisplaySafeUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self}")
+    }
+}
+
+impl From<Url> for DisplaySafeUrl {
+    fn from(url: Url) -> Self {
+        DisplaySafeUrl(url)
+    }
+}
+
+impl From<DisplaySafeUrl> for Url {
+    fn from(url: DisplaySafeUrl) -> Self {
+        url.0
+    }
+}
+
+impl FromStr for DisplaySafeUrl {
+    type Err = url::ParseError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        Ok(Self(Url::from_str(input)?))
+    }
+}
+
+fn fmt_with_obfuscated_credentials<W: Write>(url: &Url, mut f: W) -> std::fmt::Result {
+    if url.password().is_none() && url.username() == "" {
+        return write!(f, "{url}");
+    }
+
+    write!(f, "{}://", url.scheme())?;
+
+    if url.username() != "" && url.password().is_some() {
+        write!(f, "{}", url.username())?;
+        write!(f, ":****@")?;
+    } else if url.username() != "" {
+        write!(f, "****@")?;
+    } else if url.password().is_some() {
+        write!(f, ":****@")?;
+    }
+
+    write!(f, "{}", url.host_str().unwrap_or(""))?;
+
+    if let Some(port) = url.port() {
+        write!(f, ":{port}")?;
+    }
+
+    write!(f, "{}", url.path())?;
+    if let Some(query) = url.query() {
+        write!(f, "?{query}")?;
+    }
+    if let Some(fragment) = url.fragment() {
+        write!(f, "#{fragment}")?;
+    }
+
+    Ok(())
+}
+
+/// A wrapper around a [`url::Url`] ref that safely handles credentials for
+/// logging purposes.
+///
+/// Uses the same underlying [`Display`] implementation as [`DisplaySafeUrl`].
+///
+/// # Examples
+///
+/// ```
+/// use uv_redacted::DisplaySafeUrl;
+/// use std::str::FromStr;
+///
+/// // Create from a `url::Url` ref
+/// let url = Url::parse("https://user:password@example.com").unwrap();
+/// let log_safe_url = DisplaySafeUrlRef::from(&url);
+///
+/// // Display will mask secrets
+/// assert_eq!(url.to_string(), "https://user:****@example.com/");
+///
+/// // Since `DisplaySafeUrlRef` provides full access to the underlying `Url` through a
+/// // `Deref` implementation, you can still access the username and password
+/// assert_eq!(url.username(), "user");
+/// assert_eq!(url.password(), Some("password"));
+pub struct DisplaySafeUrlRef<'a>(&'a Url);
```

Comment:
This type can / should be `Copy`. It doesn't really make sense for users to use `&DisplaySafeUrlRef` since it's already a reference.

Reply:
Implemented this and your other suggestions in a follow-up PR: #13683.